File: program10.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define MAX 100

int heap[MAX];
int size = 0;

// swap
void swap(int *a, int *b) {
    int t = *a;
    *a = *b;
    *b = t;
}

// insert
void insert(int val) {
    heap[++size] = val;
    int i = size;
    while (i > 1 && heap[i / 2] > heap[i]) {
        swap(&heap[i], &heap[i / 2]);
        i /= 2;
    }
}

// heapify down
void heapify(int i) {
    int smallest = i, left = 2 * i, right = 2 * i + 1;
    if (left <= size && heap[left] < heap[smallest])
        smallest = left;
    if (right <= size && heap[right] < heap[smallest])
        smallest = right;
    if (smallest != i) {
        swap(&heap[i], &heap[smallest]);
        heapify(smallest);
    }
}

// decrease key
void decreaseKey(int i, int newVal) {
    heap[i] = newVal;
    while (i > 1 && heap[i / 2] > heap[i]) {
        swap(&heap[i], &heap[i / 2]);
        i /= 2;
    }
}

// extract minimum
int extractMin() {
    if (size == 0) return INT_MIN;
    int root = heap[1];
    heap[1] = heap[size--];
    heapify(1);
    return root;
}

// delete key
void deleteKey(int i) {
    decreaseKey(i, INT_MIN);
    extractMin();
}

// display
void display() {
    if (size == 0) {
        printf("Heap empty\n");
        return;
    }
    printf("Heap elements: ");
    for (int i = 1; i <= size; i++)
        printf("%d ", heap[i]);
    printf("\n");
}

int main() {
    int ch, val, pos;
    while (1) {
        printf("\n1.Insert\n2.Display\n3.Decrease Key\n4.Delete Key\n5.Exit\nEnter choice: ");
        scanf("%d", &ch);
        switch (ch) {
            case 1:
                printf("Enter value: ");
                scanf("%d", &val);
                insert(val);
                break;
            case 2:
                display();
                break;
            case 3:
                printf("Enter index and new value: ");
                scanf("%d%d", &pos, &val);
                decreaseKey(pos, val);
                break;
            case 4:
                printf("Enter index to delete: ");
                scanf("%d", &pos);
                deleteKey(pos);
                break;
            case 5:
                exit(0);
            default:
                printf("Invalid choice\n");
        }
    }
}

File: program10.py
# Program 10: Min Heap with Insert, Decrease Key and Delete Element

class MinHeap:
    def __init__(self):
        self.heap = []

    # Function to insert a value into the heap
    def insert(self, value):
        self.heap.append(value)
        self._heapify_up(len(self.heap) - 1)

    # Function to decrease the key at a given index
    def decrease_key(self, index, new_value):
        if index < 0 or index >= len(self.heap):
            print("Invalid index")
            return

        if new_value > self.heap[index]:
            print("New key is greater than current key, cannot decrease.")
            return

        self.heap[index] = new_value
        self._heapify_up(index)
        print(f"Key decreased at index {index} to {new_value}")

    # Function to delete element at given index
    def delete(self, index):
        if index < 0 or index >= len(self.heap):
            print("Invalid index")
            return
        
        print(f"Deleting element {self.heap[index]} at index {index}")
        self.decrease_key(index, float('-inf'))
        self.extract_min()

    # Function to remove the minimum element
    def extract_min(self):
        if not self.heap:
            print("Heap is empty.")
            return None
        
        minimum = self.heap[0]
        self.heap[0] = self.heap[-1]
        self.heap.pop()
        self._heapify_down(0)
        return minimum

    # Heapify upward (Fix on insert or decrease key)
    def _heapify_up(self, index):
        parent = (index - 1) // 2
        while index > 0 and self.heap[index] < self.heap[parent]:
            self.heap[index], self.heap[parent] = self.heap[parent], self.heap[index]
            index = parent
            parent = (index - 1) // 2

    # Heapify downward (Fix after delete / extract min)
    def _heapify_down(self, index):
        smallest = index
        left = 2 * index + 1
        right = 2 * index + 2

        if left < len(self.heap) and self.heap[left] < self.heap[smallest]:
            smallest = left
        if right < len(self.heap) and self.heap[right] < self.heap[smallest]:
            smallest = right

        if smallest != index:
            self.heap[index], self.heap[smallest] = self.heap[smallest], self.heap[index]
            self._heapify_down(smallest)

# -----------------------
# MAIN PROGRAM
# -----------------------

heap = MinHeap()

while True:
    print("\n--- Min Heap Operations ---")
    print("1. Insert")
    print("2. Decrease Key")
    print("3. Delete Element")
    print("4. Extract Min")
    print("5. Display Heap")
    print("6. Exit")

    choice = int(input("Enter your choice: "))

    if choice == 1:
        value = int(input("Enter value to insert: "))
        heap.insert(value)

    elif choice == 2:
        index = int(input("Enter index to decrease: "))
        new_value = int(input("Enter new value: "))
        heap.decrease_key(index, new_value)

    elif choice == 3:
        index = int(input("Enter index to delete: "))
        heap.delete(index)

    elif choice == 4:
        print("Extracted Min:", heap.extract_min())

    elif choice == 5:
        print("Current Heap:", heap.heap)

    elif choice == 6:
        print("Exiting...")
        break
        
    else:
        print("Invalid choice! Try again.")
